### PR TITLE
Refactor pop_verifier into gen_server

### PR DESCRIPTION
**Pull request by antonbobkov**
gen_server hype!

### CODE REVIEW

**Comment by antonbobkov** [comment]
Not very interesting, just pushing some code around

Now pop_verifier is a proper gen_server

Also build in json config file support into pop_verifier

Next step: supervisors!! 🙀

**Comment by pdlla** [comment]
I was gonna review but then started playing gwent 😳

File: test/pop_verifier_udp_test.erl [hunk 1]
```diff
@@ -106,7 +103,7 @@ udp_ver_start(NetAddress = {_IP, Port}, UdpServerName, ForwardFn, VerTot) ->
 udp_ver_stop({VerPidList, UdpServerName}) ->
 
     lists:foreach(fun(VerPid) ->
-			  VerPid ! exit
+			  gen_server:stop(VerPid)
```
Comment (pdlla) [inline test/pop_verifier_udp_test.erl]: nice!

File: test/pop_verifier_udp_test.erl [hunk 2]
```diff
@@ -52,18 +52,15 @@ start_pv(VerifierArr, PrivateKey, VerifierIndex, UdpServerId) ->
...
-				       PopManagerConfig, 
-				       PopVerifierConfig, 
-				       no_timer, 
-				       fun () -> ok end
-				      )
-			     end),
+    InitData = {explicit, PopChainConfig, PopManagerConfig, PopVerifierConfig},
+    
+    
+    {ok, VerifierPid} = gen_server:start_link(pop_verifier, InitData, []),
```
Comment (pdlla) [inline test/pop_verifier_udp_test.erl]: nice!!!

File: test/pop_verifier_json_test.erl [hunk 3]
```diff
@@ -5,82 +5,22 @@
...
-				     pop_verifier:start_loop( 
-				       PopChainConfig, 
-				       PopManagerConfig, 
-				       PopVerifierConfig, 
-				       no_timer, 
-				       fun () -> ok end
-				      )
-			     end),
+    NetSendFn = fun(DestAddressList, MsgId, Data) -> 
+			gen_server:cast({global, UdpServerId}, {send, DestAddressList, {MyAddress, MsgId, Data} })
```
Comment (pdlla) [inline test/pop_verifier_json_test.erl]: nice!!

File: src/potato_records.hrl [hunk 4]
```diff
@@ -11,14 +11,29 @@
...
 -record(pop_manager_config, {request_range_backup, 
-			     net_send, 
 			     net_multi_send, 
 			     on_new_block}).
 
 -record(pop_manager, {unbound_blocks, pop_chain, config}).
 
--record(pop_verifier_config, {sub_time_out, my_index, my_key}).
--record(pop_verifier, {config, verifiers_arr, subscribers, pop_manager}).
+-record(pop_verifier_config,
```
Comment (pdlla) [inline src/potato_records.hrl]: :O

File: src/potato_records.hrl [hunk 5]
```diff
@@ -11,14 +11,29 @@
...
+	{
+	  sub_time_out, 
+	  my_index, 
+	  my_key,
+	  event_fn,
+	  timer_interval
+	}).
+
+-record(pop_verifier,
```
Comment (pdlla) [inline src/potato_records.hrl]: reformatting +1

File: src/pop_verifier_old_code [hunk 6]
```diff
@@ -0,0 +1,181 @@
+%% @doc code that handles running a full verifier node.
```
Comment (pdlla) [inline src/pop_verifier_old_code]: just delete and put faith in version control 😱

File: src/pop_verifier.erl [hunk 7]
```diff
@@ -1,51 +1,20 @@
-%% @doc code that handles running a full verifier node.
-
 -module(pop_verifier).
+-behavior(gen_server).
```
Comment (pdlla) [inline src/pop_verifier.erl]: 🎉

File: src/pop_verifier.erl [hunk 8]
```diff
@@ -1,51 +1,20 @@
...
-    PopVerifier#pop_verifier{subscribers = maps:put(SenderAddress, CurrentTime, Subs)};
-
-on_net_message(SenderAddress, CurrentTime, MsgId, Data, PopVerifier) ->
-    PM0 = PopVerifier#pop_verifier.pop_manager,
-
-    PM1 = pop_manager:on_net_message(SenderAddress, CurrentTime, MsgId, Data, PM0),
-    
-    PopVerifier#pop_verifier{pop_manager = PM1}.
-
+make_event(Code, Data, State) ->
```
Comment (pdlla) [inline src/pop_verifier.erl]: name of this function is confusing

File: src/pop_verifier.erl [hunk 9]
```diff
@@ -101,81 +68,246 @@ on_timer(CurrentTime, PV0) ->
...
+	    CurrentTime
+    end.
+
+json_get(Key, Map) when is_atom(Key) ->
+    maps:get(atom_to_binary(Key, utf8), Map).
+
+%% json_get(Key, Map) when is_list(Key) ->
+%%     maps:get(list_to_binary(Key), Map).
+
+make_verifier_array_from_json(JsonConf) ->
```
Comment (pdlla) [inline src/pop_verifier.erl]: cool!

File: src/pop_verifier.erl [hunk 10]
```diff
@@ -101,81 +68,246 @@ on_timer(CurrentTime, PV0) ->
...
+
+    init({json_map, JsonConf, ConfigData});
+
+init({json_map, JsonConf, ConfigData}) ->
+    {PopConfigData, PopManagerConfig, PopVerConfig} = create_config_from_json(JsonConf, ConfigData),
+
+    init({explicit, PopConfigData, PopManagerConfig, PopVerConfig});
+    
+
+init(InitData = {explicit, PopConfigData, PopManagerConfig, PopVerConfig}) ->
```
Comment (pdlla) [inline src/pop_verifier.erl]: cool multiple init functions!

File: src/pop_verifier.erl (hunk 8 above)
Comment (pdlla) [reply src/pop_verifier.erl]: also comments, I guess I"m confused by this event thing in general. Why does pop_verifier need a event handler? Isn't that just suppose to be what handle_info does?

**Comment by pdlla** [comment]
woah, jsut saw the code coverage report :o

File: src/pop_verifier_old_code (hunk 6 above)
Comment (antonbobkov) [reply src/pop_verifier_old_code]: :scream:

File: src/pop_verifier.erl (hunk 8 above)
Comment (antonbobkov) [reply src/pop_verifier.erl]: not a handler, it has a way of triggering events

someone else would be in charge of processing triggered events

**Pull request merged**